### ISSUE TITLE
potential bug in PyMISP.__send_attributes()

### ISSUE THREAD

**Issue by github-germ** [MISP/MISP#3293]
I suspect there is a bug in `PyMISP.__send_attributes()` if the `__prepare_request()` does not get a `200 Response`, e.g. a `503`. In that case I believe the the line below appending to `responses` the request response converted to `json` throws an exception.

excerpt from `pymisp/api.py`
```
    def _send_attributes(self, event, attributes, proposal=False):
    .
    .
    .

    	    # __prepare_request(...) returns a requests.Response Object
            responses.append(self.__prepare_request('POST', url, json.dumps(data, cls=MISPEncode)).json())
        return responses
```

I've seen intermittent exceptions from that line on my "slow" MISP instance VM when `add_hashes()` attempts 3 hashes in the call: `md5`, `sha1`, `sha256`. Perhaps my server can't keep up and replies with a HTTP 503, triggering the `json` exception. See a suspected case below.
```
[20180525-15:00:35 ERROR (140058413758208)] add_hashes()
exception="Expecting value: line 1 column 1 (char 0)
[20180525-15:00:35 ERROR (140058413758208)] Traceback (most
recent call last):
  File "./PySight.py", line 883, in is_map_alert_to_event
    to_ids   = False)
  File "/usr/lib/python3.6/site-packages/pymisp/api.py", line
566, in add_hashes
    return self._send_attributes(event, attributes, proposal)
  File "/usr/lib/python3.6/site-packages/pymisp/api.py", line
512, in _send_attributes
    responses.append(self.__prepare_request('POST', url,
json.dumps(data, cls=MISPEncode)).json())
  File "/usr/lib/python3.6/site-packages/requests/models.py",
line 892, in json
    return complexjson.loads(self.text, **kwargs)
  File
"/usr/lib64/python3.6/site-packages/simplejson-3.13.2-py3.6-linux
-x86_64.egg/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File
"/usr/lib64/python3.6/site-packages/simplejson-3.13.2-py3.6-linux
-x86_64.egg/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File
"/usr/lib64/python3.6/site-packages/simplejson-3.13.2-py3.6-linux
-x86_64.egg/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1
column 1 (char 0)
```

Tomorrow, I will hack `api.py` locally to catch the exception with that call to `json()` and see if that helps.

**Comment by github-germ** [comment]
Pretty sure this is indeed a bug, e.g. this script emulates what `PyMISP.__prepare_request()` does (the call stack is `add_hashes()` calls `_send_attributes()` calls `__prepare_request()` and then `_send_sttributes()` throws the exception when calling `json()` on the `503 Response` from the return value from `requests.Session.send()` from `__prepare_request()`.

**Question**: Does PyMISP generally do retries, e.g. for a 503, or is that left to the client code? I.e I'm not sure of the policy to handle this exception back to the client.

```
$ cat test.503.py
#!/usr/bin/env python

import json
import requests

req = requests.Request('GET', 'http://httpstat.us/503')
s = requests.Session()
prepped = s.prepare_request(req)
prepped.headers.update({
    'Content-Type':     'application/json',
    'Accept':           'application/json',
    'User-Agent':       'test/0.0.0',})
response = s.send(prepped)
print(response)
print(response.status_code)
print(response.json())
$ 
$ python ./test.503.py
<Response [503]>
503
Traceback (most recent call last):
  File "./test.503.py", line 16, in <module>
    print(response.json())
  File "/usr/local/lib/python2.7/site-packages/requests/models.py", line 866, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/simplejson/__init__.py", line 385, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python2.7/site-packages/simplejson/decoder.py", line 402, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python2.7/site-packages/simplejson/decoder.py", line 420, in raw_decode
    raise JSONDecodeError("No JSON object could be decoded", s, idx)
simplejson.decoder.JSONDecodeError: No JSON object could be decoded: line 1 column 0 (char 0)
$ 
```

**Comment by github-germ** [comment]
duh, missed your earlier fix. that looks right. THANKS!!

**Comment by Rafiot** [comment]
Let me know if it didn't fix it. 

Note that you will get a text blob in the list, which isn't really what you want, but that's the best I can do there. 
If you can give us one of those text contents, we will be able to figure out what's failing on MISP side

**Comment by github-germ** [comment]
will do. thanks.

**Comment by github-germ** [comment]
the `text` returned is shown after the word `returned` in the log excerpt below:
```
[20180531-17:58:58 ERROR (140145830725376)] failure for reportId=18-0000\
3998 when add_hashes(event_id=11576, category="Payload delivery", filename="una\
vailable", md5="fa1937b188cbb7fd371984010564db6e", sha1="7e1679aec69b8c471f93e2\
27ec0becb5c83fe2f3", sha256="188f1ef6cb0a5f70988765d44ebf0366ffc06dcb9a4a236080\
ed5cdbbcfc5382") returned <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>503 Service Unavailable</title>
</head><body>
<h1>Service Unavailable</h1>
<p>The server is temporarily unable to service your
request due to maintenance downtime or capacity
problems. Please try again later.</p>
</body></html>

```

**Questions**

1. In all cases on `text` returned vs `json` does it mean the hash was not added (of course it might have already existed, in which case, we get a `dict` with an `error` key, and I see an HTTP 403 for that case)?

1. Might it _better_ if instead of returning a `text`, that you construct an `error dict` similar to when the hash already exists?

1. Perhaps in the case of 503, my client code can implement retry logic for `PyMISP.add_hashes`. If so, would you be able to list other HTTP error codes which should trigger retry also?

**Comment by Rafiot** [comment]
@iglocska I think this one if for you ;)

**Comment by github-germ** [comment]
Btw, I see a lot of the `add_hashes` requests generating HTTP 503 replies as in https://github.com/MISP/MISP/issues/3293#issuecomment-393624015 when I execute my multi-threaded bulk ingestion script. MISP and ingestion are running on the same host, which is a real system (i.e. not a VM) with 64GB of RAM and 32-core cpu.

This is an issue for us. Let me know if I should create a new issue to look at why this occurs on the server-side. Perhaps we need to tune something in the Apache httpd config?

Thanks.

**Comment by github-germ** [comment]
Now getting `503`'s also when instantiating `pymisp.PyMISP()` in threads. Hoping to get a handle on why the server deems itself unavailable so often. As I asked before, could this be a tuning issue?

Thanks.

```
[20180601-06:03:26 ERROR (139745870993152)] Unexpected error in MISP init: \
(<class 'pymisp.exceptions.PyMISPError'>, PyMISPError('Unable to connect to \
MISP (https://<hostname>). Please make sure the API key and the URL are correct \
(http/https is required): Expecting value: line 1 column 1 (char 0)',), \
<traceback object at 0x7f195c5623c8>) \
...
```

**Comment by Rafiot** [comment]
503 happen when MISP barf because the backend cannot handle the load, so it is a tunning issue, yes. 

If you can share the scripts you're using, that would be easier to debug on our side.

**Comment by github-germ** [comment]
Hi Raphael, I should have noted here that I created a separate issue https://github.com/MISP/MISP/issues/3305 to track this 503 situation. I was thinking this would allow this issue to be closed by focusing only on the questions I raised above in https://github.com/MISP/MISP/issues/3293#issuecomment-393624015 more specific to `PyMISP.__send_attributes()` to get that code changed/merged in, while allowing the 503 discussion to move forward separately.

**Comment by Rafiot** [comment]
Let's move it over to the other issue then.

**Comment by github-germ** [comment]
I agree.

However, it would be real helpful to me if you could reply to these prior questions that are more specific to the change you made in `PyMISP.__send_attributes()` in this issue. Thanks Raphael!!

Questions

1. As per the change you made in https://github.com/MISP/PyMISP/commit/8c4889cc690512311975dfa1554c5efb048b73bc in all cases where `text` is returned vs `json`, does it mean the hash was not added (of course it might have already existed, in which case, we get a `dict` with an error key, and I see an HTTP 403 for that case)?

1. Might it better if instead of returning `text`, that you construct an error `dict` similar to when the hash already exists?

1. Perhaps in the case of 503, my client code can implement retry logic for `PyMISP.add_hashes()`. If so, would you be able to list other HTTP error codes which should trigger retry also? --> Maybe this should move to the 503 specific issue in https://github.com/MISP/MISP/issues/3305 for a more general solution.

**Comment by Rafiot** [comment]
1. When the output isn't a json blob, yes, the request failed. MISP is supposed to fail gracefully and tell you what went wrong. In your case, it is failing because either MISP or apache barf on the amount of queries you're doing at the same time. We're still waiting for your code (in the other issue) so we can investigate further.
2. No, it's not. You should always receive a Json blob, unless apache barf, or there is a bug in MISP. It doesn't make any sense to embed the content in a fake dictionary.
3. The 503 is an apache exception. Because your code is hammering the webserver and the response you get is basically inconsistent and there is no way to figure out what worked and what didn't work at PyMISP level.

TL;DR: the issues you're having right now are not to be handled at PyMISP level and we need to see your code so we can have a rough idea of what's going on.
